### PR TITLE
feat(platform-browser): add StateTransfer API for async retrieval

### PR DESCRIPTION
**Pull request by Toxicable**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?
```
[x] Feature
```

## What is the current behavior?
You must block the bootstrapping of the app with `this.doc!.addEventListener('DOMContentLoaded', ..)` otherwise the Json might not hahve been rendered onto the DOM yet

Issue Number: N/A

## What is the new behavior?
New API's which can automatically wait for the DOM Content to be loaded before passing the payload back.

## Does this PR introduce a breaking change?
```
[] Yes
[x] No
```
It deprecates the sync methods for data retrieval

### CODE REVIEW

File: packages/platform-browser/src/browser/transfer_state.ts [hunk 1]
```diff
@@ -82,6 +91,28 @@ export function makeStateKey<T = void>(key: string): StateKey<T> {
 export class TransferState {
   private store: {[k: string]: {} | undefined} = {};
   private onSerializeCallbacks: {[k: string]: () => {} | undefined} = {};
+  private contentLoaded: Observable<any>;
+
+  constructor(
+    @Optional() @Inject(STATE_TRANSFER_USE_ASYNC) private readonly useAsync?: boolean,
+    // TODO: at v7 remove STATE_TRANSFER_USE_ASYNC and make Document non optional
```
Comment (bchuckles) [inline packages/platform-browser/src/browser/transfer_state.ts]: is comment still valid? if not remove

**Review by bchuckles — APPROVED** [review]
LGTM

File: packages/platform-browser/src/browser/transfer_state.ts [hunk 2]
```diff
@@ -162,4 +205,12 @@ export function initTransferState(doc: Document, appId: string) {
   providers: [{provide: TransferState, useFactory: initTransferState, deps: [DOCUMENT, APP_ID]}],
 })
 export class BrowserTransferStateModule {
+  static withAsyc(): ModuleWithProviders{
```
Comment (CaerusKaru) [inline packages/platform-browser/src/browser/transfer_state.ts]: /s/Asyc/Async

**Review by CaerusKaru — COMMENTED** [review]
This wasn't exactly the idea. You should make `TransferState` into an abstract class, then have `SyncTransferState` and `AsyncTransferState` and then the `withAsync` call should provide the async one. No tokens needed

File: packages/platform-browser/src/browser/transfer_state.ts [hunk 3]
```diff
@@ -82,6 +91,28 @@ export function makeStateKey<T = void>(key: string): StateKey<T> {
...
+  ) {
+    this.contentLoaded = shareReplay.call(new Observable(obs => {
+      const contentDoneLoaded = (e?: any) => {
+        obs.next();
+        obs.complete();
+      };
+      if (document.readyState == "complete" || document.readyState == "loaded") {
+        contentDoneLoaded();
+      } else {
+        const contentLoadedLisener = document.addEventListener('DOMContentLoaded', () => {
```
Comment (trotyl) [inline packages/platform-browser/src/browser/transfer_state.ts]: Typo `Lisener` in `contentLoadedLisener`

File: packages/platform-browser/src/browser/transfer_state.ts [hunk 4]
```diff
@@ -6,8 +6,12 @@
  * found in the LICENSE file at https://angular.io/license
  */
 
-import {APP_ID, Injectable, NgModule} from '@angular/core';
+import {APP_ID, Injectable, NgModule, InjectionToken, ModuleWithProviders, Optional, Inject} from '@angular/core';
 import {DOCUMENT} from '../dom/dom_tokens';
+import { fromPromise } from 'rxjs/observable/fromPromise';
+import { Observable } from 'rxjs/Observable';
+import { map } from 'rxjs/operator/map';
```
Comment (trotyl) [inline packages/platform-browser/src/browser/transfer_state.ts]: Why not pipeable `map`?

File: packages/platform-browser/src/browser/transfer_state.ts [hunk 5]
```diff
@@ -6,8 +6,12 @@
...
+import { shareReplay } from 'rxjs/operators';
```
Comment (trotyl) [inline packages/platform-browser/src/browser/transfer_state.ts]: Use full path regarding https://github.com/angular/angular-cli/issues/9069?

File: packages/platform-browser/src/browser/transfer_state.ts [hunk 6]
```diff
@@ -82,6 +91,28 @@ export function makeStateKey<T = void>(key: string): StateKey<T> {
...
   private onSerializeCallbacks: {[k: string]: () => {} | undefined} = {};
+  private contentLoaded: Observable<any>;
+
+  constructor(
+    @Optional() @Inject(STATE_TRANSFER_USE_ASYNC) private readonly useAsync?: boolean,
+    // TODO: at v7 remove STATE_TRANSFER_USE_ASYNC and make Document non optional
+    @Inject(DOCUMENT) private readonly doc?: Document
+  ) {
+    this.contentLoaded = shareReplay.call(new Observable(obs => {
+      const contentDoneLoaded = (e?: any) => {
```
Comment (trotyl) [inline packages/platform-browser/src/browser/transfer_state.ts]: Meaning of `e?: any`? Function types should be contravariant even under new `strictFunctionTypes` flag.

File: packages/platform-browser/src/browser/transfer_state.ts [hunk 7]
```diff
@@ -82,6 +91,28 @@ export function makeStateKey<T = void>(key: string): StateKey<T> {
...
+        obs.next();
+        obs.complete();
+      };
+      if (document.readyState == "complete" || document.readyState == "loaded") {
```
Comment (trotyl) [inline packages/platform-browser/src/browser/transfer_state.ts]: Better to `===`?

File: packages/platform-browser/src/browser/transfer_state.ts [hunk 8]
```diff
@@ -82,6 +91,28 @@ export function makeStateKey<T = void>(key: string): StateKey<T> {
...
+        contentDoneLoaded();
+      } else {
+        const contentLoadedLisener = document.addEventListener('DOMContentLoaded', () => {
+          contentDoneLoaded();
+        });
+        return () => this.doc!.removeEventListener('DOMContentLoaded', contentDoneLoaded);
```
Comment (trotyl) [inline packages/platform-browser/src/browser/transfer_state.ts]: Why `this.doc` here but global `document` above?

**Comment by trotyl** [comment]
@Toxicable OK, that makes sense... 😃

File: packages/platform-browser/src/browser/transfer_state.ts (hunk 4 above)
Comment (Toxicable) [reply packages/platform-browser/src/browser/transfer_state.ts]: no where else in the framework are we using pipeable operators yet, would prefer to sick to that for now.

**Comment by Toxicable** [comment]
I don't know what this means
>node_modules/@angular/platform-browser/src/private_export.d.ts(21,71): error TS2306: File '/home/travis/build/angular/angular/integration/bazel/node_modules/@angular/platform-browser/src/browser/transfer_state_async_token.d.ts' is not a module.

Am I meant to add the file I made to somewhere?

**Comment by CaerusKaru** [comment]
It's because you marked the token as internal, e2e_2 can't find it. If you add it to private_reexport with the following line, and then remove the internal flag, it should work:

`export {STATE_TRANSFER_USE_ASYNC as ɵSTATE_TRANSFER_USE_ASYNC} from './browser/transfer_state_async_token';`

File: packages/platform-browser/src/browser/transfer_state.ts [hunk 9]
```diff
@@ -81,20 +83,46 @@ export function makeStateKey<T = void>(key: string): StateKey<T> {
...
   private onSerializeCallbacks: {[k: string]: () => {} | undefined} = {};
 
   /** @internal */
-  static init(initState: {}) {
+  static init(initState: {}, initialStore?: Promise<{[k: string]: {} | undefined}>) {
     const transferState = new TransferState();
-    transferState.store = initState;
-    return transferState;
+    if (initialStore) {
+      // poor mans fromPromise
```
Comment (robwormald) [inline packages/platform-browser/src/browser/transfer_state.ts]: remove

**Comment by Toxicable** [comment]
Closing due to concerns with page flicker that this could cause

**Pull request closed without merging**